### PR TITLE
Progress Display

### PR DESCRIPTION
**Pull request by rtmf**
I added a simple but effective progress display to this applet, which I am using to image a 30GB hard drive on an ancient laptop :)

### CODE REVIEW

File: bscp [hunk 1]
```diff
@@ -121,25 +121,34 @@ def bscp(local_filename, remote_host, remote_filename, blocksize, hashname):
...
-        remote_digest_list = [io.read(hash_total.digestsize) for i in xrange(blockcount)]
-
-        for remote_digest in remote_digest_list:
+        remote_digest_list = [(i,io.read(hash_total.digestsize),
+            sys.stderr.write((''*42)+'Read remote digest [%10d/%10d]'%(i,blockcount)),sys.stderr.flush()
+            ) for i in xrange(blockcount)]
+        sys.stderr.write('\n')
+        for block_num,remote_digest,write,flush in remote_digest_list:
+            sys.stderr.write((''*42)+'Digest local block [%10d/%10d]'%(block_num,blockcount))
+            sys.stderr.flush()
```
Comment (vog) [inline bscp]: This is a bit cumbersome. Also, flush() is unnecessarily called twice. Please simplify this code, e.g. like this:

```python
for i in xrange(blockcount):
    remote_digest = io.read(hash_total.digestsize)
    sys.stderr.write(...)
    sys.stderr.write(...)
    sys.stderr.flush()
    ...

**Comment by vog** [comment]
I like the idea, but the code is a bit cumbersome. Please have a look at my code comment

**Comment by rtmf** [comment]
I have looked over your commentary, calling flush is done where it is because I want to display a message at the start of a block's hashing, and then at the start of any potential transfer, and each message requires a flush() to be certain it hits the console.  This is unlikely to be visible on a fast system and I'd like to redo this with a progress bar that shows symbols for whether the block differed and was transferred or not, or somthing, this was quick&dirty

**Comment by vog** [comment]
Maybe I'm reading the code differently, but as I understand it, the order of actions is:

* read() is called
* the message "Read remote digest" is printed and flushed
* the message "Digest local block" is printed and flushed
* the hash is calculated

Anyway, please simplify the generator/loop construction such that these details are obvious rather than convoluted.

I don't mind the progress display to be simple-minded, because having *some* progress display is better than having none.

One more thing: Please don't use special characters in the source file, i.e. replace `'^H'` (or whatever is looks like in your editor) with `'\x08'`.

Finally, feel free to add yourself to the authors list at the top of the file, and to extend the year range.

**Comment by rtmf** [comment]
yeah you are reading the code differently, the remote digests are all read first, as there is no threading going on here, and from runtime experience with a slow machine, it is clear that the entire generator is being processed first before it is iterated, no lazy evaluation here.  As for the special character for backspace, I will certainly change that, it would make much more sense to use a bit of tput or curses here and read terminfo properly rather than just presuming backspace works and we counted properly.

**Comment by vog** [comment]
rtmf wrote:

> yeah you are reading the code differently

I guess this just reinforces my point about improving code clarity. ;-)

**Comment by xtaran** [comment]
An alternative option to including code to show the progress is to use the tool [progress](https://github.com/Xfennec/progress) which also works fine with bscp, at least if you give it the according pid:

```
# progress -p 6986
[ 6986] python2.7 /dev/dm-12
        6.9% (81.1 GiB / 1.1 TiB)

# progress -M -p 6986
[ 6986] python2.7 /dev/dm-12
        7.0% (82.0 GiB / 1.1 TiB) 21.9 MiB/s remaining 14:07:35

```

**Comment by vog** [comment]
To anyone involved in this issue, please see #16 "Future of bscp".

**Comment by vog** [comment]
I'm closing and rejecting this pull request for now, as the work on cleaning up this pull request has stopped years ago.

@rtmf Please feel free to open a new pull request if you like to make a second proposal. I'd still love to see progress information in Bscp, given that the code is simple and clean.

**Pull request closed without merging**